src/python/pants/goal/migrate_call_by_name.py

```diff
@@ -613,19 +611,35 @@ def visit_AsyncFunctionDef(self, node: ast.AsyncFunctionDef):
         if not self._should_visit_node(node.decorator_list):
             return
 
-        for child in node.body:
-            if call := self._maybe_replaceable_call(child):
-                if replacement := self.syntax_mapper.map_get_to_new_syntax(
-                    call, self.filename, calling_func=node.name
-                ):
-                    self.replacements.append(replacement)
+        self._recurse_body_statements(node.name, node)
+
+    def _recurse_body_statements(self, root: str, node: ast.stmt):
+        """Recursively walk the body of a node, including properties of compound statements looking
+        for Get() calls to replace.
+
+        https://docs.python.org/3/reference/compound_stmts.html
+        """
+        for prop in ["body", "handlers", "orelse", "finalbody"]:
+            for child in getattr(node, prop, []):
+                self._recurse_body_statements(root, cast(ast.stmt, child))
 
-            for call in self._maybe_replaceable_multiget(child):
-                if replacement := self.syntax_mapper.map_get_to_new_syntax(
-                    call, self.filename, calling_func=node.name
-                ):
-                    replacement.is_argument = True
-                    self.replacements.append(replacement)
+        self._maybe_add_replacements(root, node)
+
+    def _maybe_add_replacements(self, calling_func: str, statement: ast.stmt):
+        if call := self._maybe_replaceable_call(statement):
+            if replacement := self.syntax_mapper.map_get_to_new_syntax(
+                call, self.filename, calling_func=calling_func
+            ):
+                self.replacements.append(replacement)
+
+        for call, statement_type in self._maybe_replaceable_multiget(statement):
+            if replacement := self.syntax_mapper.map_get_to_new_syntax(
+                call, self.filename, calling_func=calling_func
+            ):
+                # Fixes an issue with generators having trailing commas causing a syntax error
+                # TODO: A better solution might be to use libcst or similar, rather than an ast
+                replacement.is_argument = statement_type == "call"
```

Comment:
This comment doesn't make sense to me without the diff to the prior version. (i.e. I managed to get it in the context of this PR, but I wouldn't when encountering this comment in the source tree in the future.)

Perhaps better suited as a PR code review comment?

Reply:
That line was the motivating factor for https://github.com/pantsbuild/pants/issues/21040 - literally, that single trailing comma.

Essentially, without differentiating between calls and generators, we end up in this situation:

```python
await MultiGet(
  foo(...),
  bar(...), <-- trailing comma fine
)
```

```python
await MultiGet(
  foo(...), <-- trailing comma == syntax error
  for x in y
)
```

The actual location of the problem is during the `fileinput` replacement (the second reason for wanting `libcst` - to see if that function could be replaced). This would be the call in question: `line_end = ",\n" if replacement.is_argument else "\n"`

To replace it "correctly", I'd have to start adding even more CST-centric code, and re-factor it all.

Reply:
yea, I got the code, and I got the comment in context of what there was before, i.e. what was addressed. But looking at the fixed code, the comment doesn't make sense, as I don't see the issue then--so I don't know what part of the statement is "fixing what" (for a future me that forgot all this context). ((I think the comment could either expand a little, or just be removed, as I don't feel we need to justify lines of code with comments--the TODO is perfectly fine tho))

Reply:
Okay, re-thought it, and `is_argument` was just a stale name. Re-named it to the far more useful `add_trailing_comma` and pushed the CST-related comment into the dataclass.

So, now we know why that exists, and we know that I don't like it and it should be replaced 😆